base/strings/util.jl

```diff
@@ -195,6 +195,60 @@ end
 # TODO: optimization for the default case based on
 # chop(s::AbstractString) = SubString(s, firstindex(s), prevind(s, lastindex(s)))
 
+"""
+    chopprefix(s::AbstractString, prefix::AbstractString) -> SubString
+
+Remove the prefix `prefix` from `s`. If `s` does not start with `prefix`, a string equal to `s` is returned.
+
+See also [`chopsuffix`](@ref).
+
+!!! compat "Julia 1.7"
+    This function is available as of Julia 1.7.
+
+# Examples
+```jldoctest
+julia> chopprefix("Hamburger", "Ham")
+"burger"
+
+julia> chopprefix("Hamburger", "hotdog")
+"Hamburger"
+```
+"""
+function chopprefix(s::AbstractString, prefix::AbstractString)
+    if startswith(s, prefix)
+        SubString(s, nextind(s, lastindex(prefix), 1), lastindex(s))
+    else
+        SubString(s)
+    end
```

Comment:
I think we can do this fairly compactly as follows (`startswith` could also be written this way, to show they are the same):
```suggestion
    i, j = iterate(a), iterate(b)
    while true
        j === nothing && i === nothing && return SubString(s, 1, 0) # s == prefix: empty result
        j === nothing && return SubString(s, i[2]) # ran out of prefix: success!
        i === nothing && return SubString(s) # ran out of source: failure
        i[1] == j[1] || return SubString(s) # mismatch: failure
        i, j = iterate(a, i[2]), iterate(b, j[2])
    end
```

Then the general `chopend` is pretty much the same, but with slight changes (such as calling `iterate(Iterators.Reverse(a))` at the top, and `SubString(s, firstindex(s), i[2])` on success)